module/VuFind/src/VuFind/View/Helper/Root/ProxyUrl.php

```diff
@@ -66,10 +80,67 @@ public function __construct($config = null)
      */
     public function __invoke($url)
     {
-        $usePrefix = !isset($this->config->EZproxy->prefixLinks)
-            || $this->config->EZproxy->prefixLinks;
+        $useWebService = $this->config->EZproxy->prefixLinksWebServiceUrl ?? false;
+        if ($useWebService) {
+            $usePrefix = $this->checkUrl($url) ?? $this->checkConfig();
+        } else {
+            $usePrefix = $this->checkConfig();
+        }
+
         return ($usePrefix && isset($this->config->EZproxy->host))
             ? $this->config->EZproxy->host . '/login?qurl=' . urlencode($url)
             : $url;
     }
+
+    /**
+     * Return the configured prefixLinks setting.
+     *
+     * @return bool The configured setting, or the default
+     */
+    private function checkConfig()
+    {
+        return $this->config->EZproxy->prefixLinks ?? true;
+    }
+
+    /**
+     * Check whether the given URL requires the proxy prefix.  Cache the repsonse.
+     *
+     * @param string $url The raw URL to check
+     *
+     * @return mixed Whether the URL should be prefixed, or null if it can't be determined
+     */
+    protected function checkUrl($url)
+    {
+        $domain = parse_url($url, PHP_URL_SCHEME)
+         . '://'
+         . parse_url($url, PHP_URL_HOST);
+        $usePrefix = $this->getCachedData("proxyUrl-domainToUsePrefix-$domain");
+        if (null === $usePrefix) {
+            $usePrefix = $this->queryWebService($domain);
+            $this->putCachedData("proxyUrl-domainToUsePrefix-$domain", $usePrefix);
+        }
+        return $usePrefix;
+    }
+
+    /**
+     * Query the web service on whether to prefix URLs to a given domain.
+     *
+     * @param $domain The domain
+     *
+     * @return mixed Whether the URL should be prefixed, or null if it can't be determined
+     */
+    protected function queryWebService($domain)
+    {
+        $prefixLinksWebServiceUrl = $this->config->EZproxy->prefixLinksWebServiceUrl;
+        $queryUrl = $prefixLinksWebServiceUrl . '?url=' . $domain;
```

Comment:
Yeah I was thinking about that as a well, forgot to raise it.  @joecorall what's your preference?  Easy enough to url-encode here.

Reply:
I don't think it's necessary to urlencode. IIUC punycode allows for non-ASCII alphanumeric characters in domain names, but I just scanned all of EZProxy's supported stanzas and there are no domains that have characters outside of the character range `[^A-Za-z0-9\.\-]`. Though if ya'll feel strongly I can modify the service to require the URL passed to be percent encoded.

Reply:
I don't have a strong preference, though I tend to encode parameters habitually to avoid subtle problems. Given the nature of the data, it might make sense to set up the service so that it can accept encoded or non-encoded data for flexibility. (That is to say, if you urldecode it and it's not already urlencoded, that shouldn't hurt anything, since there aren't going to be hostnames with % characters in them anyway).

Reply:
Fair point. I'll ensure the service can handle encoded domains.

Reply:
Sounds good -- and once encoded hostnames are supported, I'd encourage @maccabeelevine to encode the values here, since that saves us from the future question of "did we forget to encode something here, or are we really not supposed to?"

Of course, we could also solve that problem by adding a comment saying "URL encoding is not required by this service" if going without encoding is preferred. :-)

Reply:
@maccabeelevine - looks like the service already handled percent encoded parameters as-is. I added a test for it, too https://github.com/lehigh-university-libraries/ezproxy-url-checker/pull/4

Reply:
I've implemented the encoding.  As a practical matter since I am passing along the scheme + domain, the "://" will end up encoded.